mmv1/products/cloudrunv2/api.yaml

```diff
@@ -973,51 +973,74 @@ objects:
-                        description: |-
-                          TCPSocket specifies an action involving a TCP port. Exactly one of HTTPGet or TCPSocket must be specified.
-                        send_empty_value: true
-                        allow_empty_object: true
-                        # exactly_one_of:
-                        #   - template.0.containers.0.livenessProbe.0.httpGet
-                        #   - template.0.containers.0.livenessProbe.0.tcpSocket
-                        properties:
-                          - !ruby/object:Api::Type::Integer
-                            name: port
-                            description: |-
-                              Port number to access on the container. Must be in the range 1 to 65535. If not specified, defaults to 8080.
+                      name: "tcpSocket"
+                      description: |-
+                        TCPSocket specifies an action involving a TCP port. Exactly one of HTTPGet or TCPSocket must be specified.
+                      send_empty_value: true
+                      allow_empty_object: true
+                      # exactly_one_of:
+                      #   - template.0.containers.0.livenessProbe.0.httpGet
+                      #   - template.0.containers.0.livenessProbe.0.tcpSocket
+                      #   - template.0.containers.0.livenessProbe.0.grpc
+                      properties:
+                        - !ruby/object:Api::Type::Integer
+                          name: port
+                          description: |-
+                            Port number to access on the container. Must be in the range 1 to 65535. If not specified, defaults to 8080.
+                    - !ruby/object:Api::Type::NestedObject
+                      name: grpc
+                      description: |-
+                        GRPC specifies an action involving a GRPC port.
+                      send_empty_value: true
+                      allow_empty_object: true
+                      # exactly_one_of:
+                      #   - template.0.containers.0.livenessProbe.0.httpGet
+                      #   - template.0.containers.0.livenessProbe.0.tcpSocket
+                      #   - template.0.containers.0.livenessProbe.0.grpc
+                      properties:
+                        - !ruby/object:Api::Type::Integer
+                          name: port
+                          description: |-
+                            Port number to access on the container. Number must be in the range 1 to 65535.
+                        - !ruby/object:Api::Type::String
+                          name: service
+                          description: |-
+                            The name of the service to place in the gRPC HealthCheckRequest
+                            (see https://github.com/grpc/grpc/blob/master/doc/health-checking.md).
+                            If this is not specified, the default behavior is defined by gRPC.
```

Comment:
Can one of these values be required.. having empty blocks in terraform code to perform behavior is confusing for customers.

Reply:
If the policy of the Terraform Google Provider is not just simply wrap the Google Cloud SDK and replicate its original behavior precisely as possible but reduce inconvenience and awkward behaviors by wrapping it with easier interface, then I would go mark the `port` property as required.
However, in that case, it would be better to also change the following properties from optional to required so that users of this resource will not get confused:

- `template.containers.startupProbe.httpGet.path`
- `template.containers.startupProbe.tcpSocket.port`
- `template.containers.livenessProbe.httpGet.path`
- `template.containers.livenessProbe.tcpSocket.port`

Reply:
Yep! We like to reduce inconvenience by making enforcing terraform blocks to not be empty by design..

If we can do this and the experience isn't negatively affected. Lets go for it. Although if it will be a net negative impact or there are reasons to the contrary for these particular fileds then please let me know!

Reply:
It allows empty blocks for tcpSocket and httpGet.

In cloudrun v1 Terraform, it also allows empty blocks: https://github.com/GoogleCloudPlatform/magic-modules/blob/main/mmv1/products/cloudrun/api.yaml#L711

Shall we keep consistency?

Reply:
It seems that there are differing preferences among the maintainers regarding to what extent Terraform Google Provider should mirror the actual behavior of the Google Cloud APIs - Terraform should precisely replicate every bit of the behaviors of the Google Cloud APIs and keep consistency with other parts of this repository, or it should work more as a user-friendly wrapper of the APIs consistently with the existing convention of Terraform providers.

Would it be possible to move this discussion topic to an issue ticket and make this pull request settled in a temporary middle ground?
It is beyond the scope of this pull request, and there will be lots of code in this repository that is relevant to this coding policy.
By moving the discussion to an issue ticket, we will be able to collect opinions from other developers as well.

Reply:
We generally would avoid allowing users sending empty blocks as it always causes more problems and some of them are hard to fix. However, if I remember correctly, based on some discussion in other PRs or offline, sending empty blocks like `tcp_socket{}` and `http_get{}` is considered different from not sending the field at all in this case. That's probably why we allow empty blocked in the first place for v1 resource, and v2 does not change the behavior (please correct me if that's not the case), so I just keep the consistency when implementing it.

@ScottSuarez do you know if we should allow users to send empty blocks if sending them is different than unset of the field?

Reply:
Yeah, one and only one of `tcp_socket`, `http_get` and ` `startupProbe` must be set in `startup_probe`. One and only one of `http_get` and ` `startupProbe` must be set in `liveness_probe`. Otherwise Cloud Run API will throw an error.